doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -473,82 +473,112 @@ Induction
      The identifiers :n:`@ident` (including the first one before the `with` clause)
      are the names of the co-induction hypotheses.
 
-.. tacn:: discriminate @term
-   :name: discriminate
-
-   This tactic proves any goal from an assumption stating that two
-   structurally different :n:`@term`\s of an inductive set are equal. For
-   example, from :g:`(S (S O))=(S O)` we can derive by absurdity any
-   proposition.
-
-   The argument :n:`@term` is assumed to be a proof of a statement of
-   conclusion :n:`@term = @term` with the two terms being elements of an
-   inductive set. To build the proof, the tactic traverses the normal forms
-   [1]_ of the terms looking for a couple of subterms :g:`u` and :g:`w` (:g:`u`
-   subterm of the normal form of :n:`@term` and :g:`w` subterm of the normal
-   form of :n:`@term`), placed at the same positions and whose head symbols are
-   two different constructors. If such a couple of subterms exists, then the
-   proof of the current goal is completed, otherwise the tactic fails.
+.. _equality-inductive_types:
 
-.. note::
-   The syntax :n:`discriminate @ident` can be used to refer to a hypothesis
-   quantified in the goal. In this case, the quantified hypothesis whose name is
-   :n:`@ident` is first introduced in the local context using
-   :n:`intros until @ident`.
+Equality of inductive types
+---------------------------
 
-.. exn:: No primitive equality found.
-   :undocumented:
+We describe in this section some special purpose tactics dealing with
+equality of inductive sets or types. These tactics use the
+equality :g:`eq:forall (A:Type), A -> A -> Prop`, simply written with the infix
+symbol :g:`=`.
 
-.. exn:: Not a discriminable equality.
-   :undocumented:
+:tacn:`discriminate` and :tacn:`injection`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. tacn:: discriminate {? @induction_arg }
+
+   .. todo PR Did I drop too much? (i.e. "The argument :n:`@one_term`
+      is assumed to be a proof of a statement of :n:`@term__1 = @term__2`"
+      How would that work with a non-trivial one_term?
```

Comment:
Yes, it would make sense that if an argument is provided, it should be a proof of an equality (typically the name of a hypothesis). [Although in the general case, it is completely possible to pass a proof term as argument instead.] You can add this to the paragraph saying that if no argument is provided, then the tactic examines all the hypotheses of the context.

Reply:
The description doesn't cover this case:

```
Goal True -> 1 <> 2.
intros.
discriminate.
```
in which the `True` is irrelevant and I expect `discriminate` only needs looks at the conclusion.

Can you give me an example in which a specific hypothesis is needed?

Also odd that this fails:

```
Goal True -> 1 = 2.
intros.
discriminate.   (* No primitive equality found. *)
```

> it should be a proof of an equality (typically the name of a hypothesis).

Is "proof of an equality" a precise statement?  That suggests that every hypothesis is a proof of its type, which would be problematic for `H: False`, eh?  What is more accurate wording?  It could be the name of lemma?

> It seems like you also forgot to mention that if the goal is of the form `term <> term`, then it behaves like `intro ident; discriminate ident` (quite similar to what happens for `injection` without any argument).

Forgot? You must be referring to Plato's theory of [anamnesis](https://en.wikipedia.org/wiki/Anamnesis_(philosophy)):

> Socrates' response is to develop his theory of anamnesis and to suggest that the soul is immortal, and repeatedly incarnated; knowledge is in the soul from eternity (86b), but each time the soul is incarnated its knowledge is forgotten in the trauma of birth.

Also not exactly like `intro ident; discriminate ident`:

```
Goal True -> 1 <> 2.
intros X.
discriminate X.  (* No primitive equality found. *)
```

So maybe should say `discriminate ...` always checks the goal.  If no `induction_arg` is given, it also checks all hypotheses.  If `induction_arg` is given, it also checks that single hypothesis or "equality proof".

Reply:
> The description doesn't cover this case:
> 
> ```
> Goal True -> 1 <> 2.
> intros.
> discriminate.
> ```

It used to cover this case but you didn't keep it in your new version. That's why I said "forgot". See the second part of the description of the last `discriminate` variant: https://coq.github.io/doc/v8.14/refman/proofs/writing-proofs/reasoning-inductives.html#id5

> Can you give me an example in which a specific hypothesis is needed?

It's never needed because if you don't provide one, Coq will look for one in your context. The only case where providing an argument would really be needed is if you want to discriminate on a proof term that you are building rather than something which is in your context (that wouldn't be common, and I expect this to be done only in proof automation). Example:

```coq
Require Import Arith.

Goal forall n, ~S n <= 0.
intros n H.
(* This is not a real example because this would also be solved immediately like this:
inversion H.
*)
discriminate (le_n_0_eq _ H).
```

> Is "proof of an equality" a precise statement?

Yes, that's a precise statement. An equality is a proposition and a type and a proof of an equality is a term whose type is an equality.

>That suggests that every hypothesis is a proof of its type, which would be problematic for `H: False`, eh?

That's actually entirely correct and not problematic at all. If you have `H: False` in your context, then, since you have a proof of `False`, you can prove your goal whatever it is.

> Also not exactly like `intro ident; discriminate ident`:
> 
> ```
> Goal True -> 1 <> 2.
> intros X.
> discriminate X.  (* No primitive equality found. *)
> ```

That's because there's one more `intro` to do. Try this instead:

```coq
Goal 1 <> 2.
intro X.
discriminate X.
(* works and is the same as calling discriminate in the first place *)
```

> So maybe should say `discriminate ...` always checks the goal. If no `induction_arg` is given, it also checks all hypotheses.

That's completely backward. It only checks the goal if no argument is given. In this case, it both checks all hypotheses and the goal if it has the form `_ <> _`.

Reply:
> >That suggests that every hypothesis is a proof of its type, which would be problematic for `H: False`, eh?

> That's actually entirely correct and not problematic at all. If you have `H: False` in your context, then, since you have a proof of `False`, you can prove your goal whatever it is.

Then the wording "proof of an equality" is incomplete.  The hypothesis is only valid within its context, not globally.  It is not a freestanding proof; that would obviously be disastrous.  It's a basic point the doc should cover somewhere.  Also, where can you find the underlying term associated with the hypothesis?

Reply:
> The hypothesis is only valid within its context, not globally.

Right. Every time the term "proof" is used in the context of a proof, this is implicitly a proof in the proof context, rather than in the global context. The proof term is most often just a name of a hypothesis.